### PR TITLE
Add support for multiplayer games

### PR DESCRIPTION
**Pull request by jonathan-marmelab**
This closes the following user story : https://trello.com/c/qdyRMGIx/8-5-as-robert-i-want-to-play-against-other-players-2

### CODE REVIEW

File: internal/model/board.go [hunk 1]
```diff
@@ -41,6 +41,13 @@ type Board struct {
 	// Players holds the players that are part of the game.
 	Players []*Player `json:"players"`
 
+	// RemainingPlayers holds the players that did not got all their targets yet.
+	RemainingPlayers []int `json:"remainingPlayers"`
+
+	// RemainingPlayerIndex is the index of the current player in the remaining
+	// players array.
+	RemainingPlayerIndex int `json:"currentPlayerIndex"`
```
[the remark this answers is not in the document]
Comment (jonathan-marmelab) [reply internal/model/board.go:49]: I named it that way since it is bound to ther remaing players array.

**Review by arimet — APPROVED** [review]

File: internal/model/board.go [hunk 2]
```diff
@@ -302,13 +325,23 @@ func randomRotation() Rotation {
 }
 
 // NewBoard returns a board for the given size.
-// The size param MUST be an odd number.
-func NewBoard(size int) (*Board, error) {
-	if (size & 1) != 1 {
-		return nil, fmt.Errorf("size must be an odd number, got: %d", size)
+// The size param MUST be either 3 or 7.
```
Comment (fzaninotto) [inline internal/model/board.go:328]: useless comment as you have the input validation below

File: internal/model/board.go [hunk 3]
```diff
@@ -319,17 +352,56 @@ func NewBoard(size int) (*Board, error) {
...
-		},
-		State: GameStatePlaceTile,
+		Tiles:                make([][]*BoardTile, size),
+		Players:              make([]*Player, 0, playerCount),
+		RemainingPlayers:     make([]int, 0, playerCount),
+		RemainingPlayerIndex: 0,
+		State:                GameStatePlaceTile,
+	}
+
+	board.Players = append(board.Players, &Player{
```
Comment (fzaninotto) [inline internal/model/board.go:362]: couldn't you represent all that logic into one data object, and loop over it? In JS, I'd go

```js
const playerPositions = [
  { color: 'blue', x: 1, y: 1 },
  { color: 'green', x: 1, y: -1 },
  // etc
]
```

This makes a data structure emerge and removes code duplication.

File: internal/model/board_test.go [hunk 4]
```diff
@@ -660,17 +805,119 @@ func TestNewBoard(t *testing.T) {
...
+			board, _ := NewBoard(3, 1)
 			assert.Equal(t, 1, len(board.Players))
 			assert.Equal(t, ColorBlue, board.Players[0].Color)
 			assert.Equal(t, 0, board.Players[0].Line)
 			assert.Equal(t, 0, board.Players[0].Row)
+			assert.Equal(t, 0, board.Players[0].Row)
+		}
+
+		{
+			board, _ := NewBoard(3, 2)
```
Comment (fzaninotto) [inline internal/model/board_test.go:817]: very hard to read

File: internal/model/player.go [hunk 5]
```diff
@@ -3,7 +3,7 @@ package model
 type Color int
 
 const (
-	ColorBlue Color = 1 << iota
```
Comment (fzaninotto) [inline internal/model/player.go:6]: Glad you removed that. Binary operations are often a bit too smart.

**Review by fzaninotto — CHANGES_REQUESTED** [review]

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**